compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESimplifyDmaWaits.cpp

```diff
@@ -0,0 +1,158 @@
+              maybeConnectionOp->emitOpError()
+                  << "expected to operate on an `amdaie.flow`";
+              return WalkResult::interrupt();
+            }
+            if (maybeFlowOp->getIsPacketFlow()) return WalkResult::advance();
+            // Retrieve the BD ID op.
+            std::optional<AMDAIE::BdIdOp> maybeBdIdOp =
+                npuHalfDmaCpyNdOp.getBdIdOp();
+            if (!maybeBdIdOp) {
+              npuHalfDmaCpyNdOp.emitOpError()
+                  << "must have a BD ID op to lower to "
+                     "`amdaie.npu.write_bd`";
+              return WalkResult::interrupt();
+            }
+            AMDAIE::BdIdOp bdIdOp = maybeBdIdOp.value();
+            // Retrieve the tile op.
+            AMDAIE::TileOp tileOp = dyn_cast_if_present<AMDAIE::TileOp>(
+                bdIdOp.getTile().getDefiningOp());
+            if (!tileOp) {
+              bdIdOp.emitOpError() << "must operate on an `amdaie.tile`";
+              return WalkResult::interrupt();
+            }
+            // Get the maximum queue size.
+            uint32_t col = getConstantIndexOrAssert(tileOp.getCol());
+            uint32_t row = getConstantIndexOrAssert(tileOp.getRow());
+            uint32_t maxQueueSize = deviceModel.getDmaMaxQueueSize(col, row);
+            // Keep wait op if reaches the maximum queue size or there is a
+            // duplicate BD ID.
+            uint32_t bdId = getConstantIndexOrAssert(bdIdOp.getValue());
+            auto &bdIdQueue = connectionToBdIdQueues[connectionOp];
+            if (bdIdQueue.size() >= maxQueueSize) bdIdQueue.clear();
+            if (bdIdQueue.empty() || llvm::is_contained(bdIdQueue, bdId)) {
+              toErase = false;
+              bdIdQueue = {bdId};
+            } else {
+              bdIdQueue.push_back(bdId);
+            }
+          }
+        }
+        // Erase later to avoid invalidating the iterator.
+        if (toErase) waitOpsToErase.push_back(waitOp);
+        return WalkResult::advance();
+      });
+  if (res.wasInterrupted()) return failure();
+
+  for (AMDAIE::NpuDmaWaitOp waitOp : waitOpsToErase) {
+    SmallVector<Value> asyncTokens(waitOp.getAsyncTokens());
+    // Erase the wait op.
+    rewriter.eraseOp(waitOp);
+    for (Value token : asyncTokens) {
+      if (auto op = dyn_cast_if_present<AMDAIE::NpuHalfDmaCpyNdOp>(
+              token.getDefiningOp())) {
+        if (op.use_empty()) {
+          rewriter.setInsertionPoint(op);
+          TypeRange resultTypeRange = TypeRange{};
+          // Nullify the result to avoid issuing a token.
+          rewriter.create<AMDAIE::NpuHalfDmaCpyNdOp>(
+              op.getLoc(), resultTypeRange, op.getConnection(), op.getInput(),
+              op.getMixedOffsets(), op.getMixedSizes(), op.getMixedStrides(),
+              op.getBdId(), op.getChannel());
+          rewriter.eraseOp(op);
+        }
+      }
+    }
+  }
+
+  return success();
+}
+
+class AMDAIESimplifyDmaWaitsPass
```

Comment:
`Simplify` is a quite broad term. I think for example `AMDAIEFoldDmaWaitsPass` is more descriptive.